dap: make exec shell persistent across the build

jsternberg opened the pull request:
Invoking the shell will cause it to persist across the entire build and
to re-execute whenever the builder pauses at another location again.

This still requires using `exec` to launch the shell. Launching by frame
id is also removed since it no longer applies to this version.

ktock reviewed (approved):
Thanks

rcjsuen commented on dap/debug_shell.go:
```diff
@@ -0,0 +1,211 @@
+	}
+	defer ctr.Cancel()
+
+	fmt.Fprintf(in.Stdout, "Running %s in build container.\n", strings.Join(append(cfg.Entrypoint, cfg.Cmd...), " "))
```
```
Running /bin/sh in build container.
/ # 
Running /bin/sh in build container.
/ # 
Running /bin/sh in build container.
/ # 
```
Is there a way for us to include the line number or some other additional context information here I wonder? 🤔 For someone new to the feature they may think it is misbehaving if they see the same log message being sent multiple times.

crazy-max replied:
And if it's the default shell then just the following would be fine:

```
Running shell from "RUN echo foo"
/ # 
Running shell from "RUN echo bar"
/ # 
```

crazy-max replied:
> to include the line number

Or maybe line number is better as a similar command could exist in multiple places or could be very long in some Dockerfiles.

jsternberg replied:
I've included the line number and an additional message to say the changes will be reset when the build continues.

jsternberg commented:
There's a bit of a weird thing that happens with this PR that didn't for the original version and I have an idea to fix it, but it might be difficult.

At the current moment, it's not possible to cancel the exec without terminating the build. And if you exit the container process from within the terminal, it only closes the container and doesn't close the attachment which will just hang until a new process comes up.

I think it might be good to differentiate these. If the container is the one who exits before the program is resumed, I think we should terminate the connection to the client so the attach command exits. If the program is resumed and the reason for the terminal dying is because the container was canceled, we should restore it on the next breakpoint. Basically, try to differentiate from a user initiated exit and a debugger initiated exit.

Might be difficult to do but I'll give it a try. I think it'll likely make a better user experience.

jsternberg commented:
I've made some changes to this. If the exec exits with a zero status, the exec is considered "complete" and it will release control of the attached shell. If the exec exits with a non-zero status or exits because it was canceled (such as resuming the build), it will restore itself after the next pause.

I've also changed it so the exec can be done again. If you try and exec while a shell is still attached, it will give the same error. If the previous shell exited and you do exec again, it will re-use the socket and connect again.

I also moved some of the logging messages to somewhere more appropriate. I added a "reason" to the message when it doesn't invoke a build container for a paused step. This should be an improved workflow.

jsternberg commented:
Testing this out more and it might not work properly when execing on an error state. Moving this to draft for now.

jsternberg commented:
This is ready for review. The pausing on exception was broken on something unrelated to this so I'll fix that separately.

jsternberg commented:
Fixed. This is because a paused a build step was only attempting to attach to the shell once. So when you exit and try to re-enter it didn't get the shell again.

I've now modified it so it attempts to reattach on a clean exit and it won't reattach on an exit with an error which includes the context being canceled.

rcjsuen commented on dap/debug_shell.go:
```diff
@@ -0,0 +1,308 @@
+	}
+
+	cfg := &build.InvokeConfig{Tty: true}
+	if len(cfg.Entrypoint) == 0 && len(cfg.Cmd) == 0 {
+		cfg.Entrypoint = []string{"/bin/sh"} // launch shell by default
+		cfg.Cmd = []string{}
```
That would presumably happen for the WCOW case also I suppose.

jsternberg replied:
It's supposed to wait. I must have broken that with the last change.

jsternberg replied:
@crazy-max this is the output I get when trying to exec on a scratch container.

```
Attached to build process.
Build container is not executable. (reason: stat error: lstat bin/sh: no such file or directory)
```

This should prevent trying to run a container when the shell doesn't exist.

jsternberg replied:
This should now be fixed. There was an error in `StatFile` where it would attempt to use the ref from the scratch volume and it resulted in a nil reference. Now it checks for that so the `canInvoke` method works properly without crashing.
